PubNub/Data/Models/PNPAMToken.m

```diff
@@ -0,0 +1,378 @@
+        [permissions addObject:@"join"];
+    }
+    
+    return [permissions componentsJoinedByString:@", "];
+}
+
+#pragma mark -
+
+
+@end
+
+
+@implementation PNPAMTokenResource
+
+
+#pragma mark - Initialization & Configuration
+
+- (instancetype)initWithResources:(NSDictionary<NSString *, PAMResourcesDictionary *> *)resourcesPermission {
+    if ((self = [super init])) {
+        _channels = [self resourcePermissionsFrom:resourcesPermission[@"chan"]];
+        _groups = [self resourcePermissionsFrom:resourcesPermission[@"grp"]];
+        _uuids = [self resourcePermissionsFrom:resourcesPermission[@"uuid"]];
+    }
+    
+    return self;
+}
+
+- (NSDictionary<NSString *, PNPAMResourcePermission *> *)resourcePermissionsFrom:(PAMResourcesDictionary *)permissions {
+    NSMutableDictionary *resourcePermissions = [NSMutableDictionary new];
+    
+    [permissions enumerateKeysAndObjectsUsingBlock:^(NSString *identifier, NSNumber *accessBits, BOOL *stop) {
+        resourcePermissions[identifier] = [PNPAMResourcePermission permissionWithValue:accessBits.unsignedIntegerValue];
+    }];
+    
+    return resourcePermissions;
+}
+
+
+#pragma mark - Misc
+
+- (NSString *)description {
+    NSMutableString *channels = [NSMutableString new];
+    NSMutableString *groups = [NSMutableString new];
+    NSMutableString *uuids = [NSMutableString new];
+    
+    [self.channels enumerateKeysAndObjectsUsingBlock:^(NSString *channel, PNPAMResourcePermission *permission, BOOL *stop) {
+        [channels appendFormat:@"\n\t\t\t- %@: %@", channel, permission.description];
+    }];
+    
+    [self.groups enumerateKeysAndObjectsUsingBlock:^(NSString *group, PNPAMResourcePermission *permission, BOOL *stop) {
+        [groups appendFormat:@"\n\t\t\t- %@: %@", group, permission.description];
+    }];
+    
+    [self.uuids enumerateKeysAndObjectsUsingBlock:^(NSString *uuid, PNPAMResourcePermission *permission, BOOL *stop) {
+        [uuids appendFormat:@"\n\t\t\t- %@: %@", uuid, permission.description];
+    }];
+    
+    return [NSString stringWithFormat:@"\n\t\t- channels: %@"
+            "\n\t\t- channel groups: %@"
+            "\n\t\t- uuids: %@",
+            channels, groups, uuids];
+}
+
+
+#pragma mark -
+
+
+@end
+
+
+@implementation PNPAMToken
+
+
+#pragma mark - Information
+
+- (NSUInteger)version {
+    return ((NSNumber *)self.parsedToken[@"v"]).unsignedIntegerValue;
+}
+
+- (NSUInteger)timestamp {
+    return ((NSNumber *)self.parsedToken[@"t"]).unsignedIntegerValue;
+}
+
+- (NSUInteger)ttl {
+    return ((NSNumber *)self.parsedToken[@"ttl"]).unsignedIntegerValue;
+}
+
+- (NSString *)authorizedUUID {
+    return self.parsedToken[@"uuid"];
+}
+
+
+- (NSDictionary *)meta {
+    return self.parsedToken[@"meta"];
+}
+
+- (NSData *)signature {
+    return self.parsedToken[@"sig"];
+}
+
+
+#pragma mark - Initialization & Configuration
+
++ (instancetype)tokenFromBase64String:(NSString *)string {
+    return [[self alloc] initFromBase64String:string];
```

Comment:
sorry for asking, but I'm not much familoar with Obj-C, how to ensure that it gets deallocated eventually?

Reply:
@bartk Objective-C uses automatic reference counting.. as soon as last strong reference on object is gone, garbage collector will release allocated memory.

Reply:
ok, I bet you right ;) Good to know that.